src/groth16/aggregate/accumulator.rs

```diff
@@ -196,29 +202,35 @@ where
             // not one since 1^r = 1
             outt = outt.pow(&coeff.into_repr());
         }
-        PairingCheck(miller_out, outt)
+        PairingCheck {
+            left: miller_out,
+            right: outt,
+            randomized: true,
+        }
     }
 
     /// takes another pairing tuple and combine both sides together as a random
     /// linear combination.
     fn merge(&mut self, p2: &PairingCheck<E>) {
         // multiply miller loop results together
-        self.0.mul_assign(&p2.0);
+        self.left.mul_assign(&p2.left);
         // multiply right side in GT together
-        if p2.1 != E::Fqk::one() {
-            if self.1 != E::Fqk::one() {
+        if p2.right != E::Fqk::one() {
+            if self.right != E::Fqk::one() {
                 // if both sides are not one, then multiply
-                self.1.mul_assign(&p2.1);
+                self.right.mul_assign(&p2.right);
             } else {
                 // otherwise, only keep the side which is not one
-                self.1 = p2.1.clone();
+                self.right = p2.right.clone();
             }
         }
+        // A merged PairingCheck is only randomized if both of its contributors are.
+        self.randomized = self.randomized && p2.randomized;
```

Comment:
Im not sure I understand the logic here.
When you merge two "native" pairing checks together, if one of them is randomized, then you can already verify. this is a bit counter intuitive here.